scanner/scanner.go

```diff
@@ -344,54 +266,32 @@ func (s *Scanner) Scan(ctx context.Context, foundCert func(*ct.LogEntry), foundP
 		close(stop)
 	}()
 
-	fetches := make(chan fetchRange)
+	batches := make(chan EntryBatch) // Output from the Fetcher.
 	jobs := make(chan entryInfo, s.opts.BufferSize)
+	go func() { // Flatten the output from Fetcher jobs.
+		defer close(jobs)
+		for b := range batches {
+			for i, e := range b.entries {
+				jobs <- entryInfo{index: b.start + int64(i), entry: e}
+			}
+		}
+	}()
 
-	var ranges list.List
-	for start, end := s.opts.StartIndex, s.opts.EndIndex; start < end; {
-		batchEnd := min(start+int64(s.opts.BatchSize), end)
-		ranges.PushBack(fetchRange{start, batchEnd - 1})
-		start = batchEnd
-	}
-	var fetcherWG sync.WaitGroup
-	var matcherWG sync.WaitGroup
-	// Start matcher workers
-	for w := 0; w < s.opts.NumWorkers; w++ {
-		matcherWG.Add(1)
-		go func(w int) {
-			defer matcherWG.Done()
+	// Start matcher workers.
+	var wg sync.WaitGroup
+	for w, cnt := 0, s.opts.NumWorkers; w < cnt; w++ {
+		wg.Add(1)
+		go func(idx int) {
+			defer wg.Done()
 			s.matcherJob(jobs, foundCert, foundPrecert)
-			s.Log(fmt.Sprintf("Matcher %d finished", w))
+			s.Log(fmt.Sprintf("Matcher %d finished", idx))
 		}(w)
 	}
-	// Start fetcher workers
-	for w := 0; w < s.opts.ParallelFetch; w++ {
-		fetcherWG.Add(1)
-		go func(w int) {
-			defer fetcherWG.Done()
-			s.fetcherJob(ctx, fetches, jobs)
-			s.Log(fmt.Sprintf("Fetcher %d finished", w))
-		}(w)
+	if err := s.fetcher.Run(ctx, batches); err != nil {
+		return err // FIXME: What to do with wg?
```

Comment:
Prefer TODO(who) / TODO(issue#) format

Reply:
Could have a local cancel for the context and call that, then `wg.Wait`?

Reply:
Yeah, I just forgot about this FIXME. How about closing the channel and waiting workers to terminate?